toc/rfc/rfc-draft-ssh-over-ws.md

```diff
@@ -0,0 +1,329 @@
+
+## Problem
+
+Cloud Foundry is a shared multi-tenant platform.
+
+If a feature requiring an open port such as CF SSH is used by at least one customer, the port has to be opened on the whole platform. It cannot be enabled for one tenant but disabled for another.
+
+In production deployments of Cloud Foundry, it is commonplace for customers to run security scans on their workloads to ensure compliance and reduce attack surfaces.
+
+These scans usually complain that besides port 443 for HTTP traffic, also port 2222 is open for SSH on the internet-facing load balancer. Many customers demand closure of the port as they
+view it as a potential attack surface for brute-force attacks against the SSH protocol.
+
+The issue is exacerbated by features like "bring-your-own-domain" where customers provide a CNAME entry alongside a matching x509 certificate that points their own (sub-)domain to an app hosted on the shared Cloud Foundry environment.
+
+This can cause potential reputation damage to the customer because to an outsider, it looks like the port 2222 is open on the customer's domain, e.g. `www.big-corp.com:2222` open looks much worse than `my-app.cfapps.eu10.hana.ondemand.com:2222` even though both are technically the same.
+
+Security-wise it is easier for an attacker to probe the environment by simply doing a port scan and then testing direct attacks on the SSH proxy:
+
+```
+nmap api.cf.eu10.hana.ondemand.com
+
+Starting Nmap 7.94 ( https://nmap.org ) at 2024-04-08 10:29 CEST
+Nmap scan report for api.cf.eu10.hana.ondemand.com (3.124.222.77)
+Host is up (0.031s latency).
+Other addresses for api.cf.eu10.hana.ondemand.com (not scanned): 3.122.209.241 3.124.208.223
+rDNS record for 3.124.222.77: ec2-3-124-222-77.eu-central-1.compute.amazonaws.com
+Not shown: 997 filtered tcp ports (no-response)
+PORT     STATE SERVICE
+80/tcp   open  http
+443/tcp  open  https
+2222/tcp open  EtherNetIP-1
+```
+
+then proceed with connecting to the port 2222:
+
+```
+nc api.cf.eu10.hana.ondemand.com 2222
+
+SSH-2.0-diego-ssh-proxy
+```
+
+the SSH proxy greets with the SSH-2.0 version, allowing the attacker to try all kinds of SSH-2.0 related probes.
+
+Another facet of the problem is that many customers require explicit allow-listing of external web sites from their on-prem environments.
+Port 443 is usually not a problem as many firewalls open it by default, but port 2222 likely is frowned upon and cause for concern.
+
+## Proposal
+
+### Note on Security
+The proposal does not aim to solve security issues of (CF) SSH per se. Tunneling a protocol through another does not make it inherently more secure. Any weakness of SSH within the CF SSH proxy will still be present even if the RFC is accepted. The main goal is to reduce the number of open ports required. Although, there is one minor security improvement as the SSH traffic will then be also encrypted by TLS within the WebSocket tunnel.
```

Comment:
If I understand this correctly, this proposal doesn't solve the problem mentioned above. Should the problem statement be something else to reflect what the proposal is solving?

Reply:
@winkingturtle-vmw It does improve on the SSH port sticking out and waiting to be exploited. I've added this section to manage the expectation that somehow removing the port will make SSH more secure, which it does not. If there is an attack possible on SSH the same attack will work on SSH over WebSockets. 
But I do get your point, any proposal for a change in the problem statement?

Reply:
I think the problem you are trying to solve has to do with SSH protocol being discoverable. With this proposal, we are hiding away that port from being discoverable.

Reply:
thank you. I've added a statement on both the problem as well as the solution sections to emphasize that.